README.rst
.. image:: https://raw.githubusercontent.com/munterfi/zeitsprung/master/docs/_static/logo.svg
   :width: 120 px
   :alt: zeitsprung.fm
   :align: right

==========
zeitsprung
==========

.. image:: https://img.shields.io/pypi/v/zeitsprung.svg
        :target: https://pypi.python.org/pypi/zeitsprung

.. image:: https://github.com/munterfi/zeitsprung/workflows/build/badge.svg
        :target: https://github.com/munterfi/zeitsprung/actions?query=workflow%3Abuild

.. image:: https://readthedocs.org/projects/zeitsprung/badge/?version=latest
        :target: https://zeitsprung.readthedocs.io/en/latest/?badge=latest
        :alt: Documentation Status

.. image:: https://pyup.io/repos/github/munterfi/zeitsprung/shield.svg
        :target: https://pyup.io/repos/github/munterfi/zeitsprung/
        :alt: Updates

.. image:: https://codecov.io/gh/munterfi/zeitsprung/branch/master/graph/badge.svg
        :target: https://codecov.io/gh/munterfi/zeitsprung

Note: zeitsprung.fm has moved to geschichte.fm, therefore this project is no longer maintained.

This package provides a scraper for www.zeitsprung.fm, a great history podcast.
To get the metadata of all episodes from the website, simply start the scraper::

    from zeitsprung.scraping import Scraper
    s = Scraper('path/to/folder/for/database')
    s.run()

The scraper then downloads the all episode metadata and audio files. The metadata is written to the 'meta' table in the
database. The audio files are converted to '.wav' files and saved separately to a folder, while a link to the file is
stored in the 'audio' table in the database.

To access the data, create a SQLiteEngine::

    from zeitsprung.database import SQLiteEngine
    db = SQLiteEngine('path/to/folder/for/database/zeitsprung.db')

Query the meta data from the database::

    db.query_all_meta()

And the audio file paths and meta data::

    db.query_all_audio()

Now have fun with analysing the episodes of zeitsprung!

Features
--------

* Scraper class to download the meta data and audio files of all episodes.
* Database class to setup and access the SQLite database containing the meta data of the episodes.

To Do
-----

* Processing class to conduct speech recognition on the audio files and build an index for clustering the topics.
* Visualize up to date statistics.

References
----------

* https://www.zeitsprung.fm, check it out!
* This package is licensed under MIT, see the LICENSE file for details.


HISTORY.rst
=======
History
=======

0.1.1 (2021-10-02)
------------------
* Adjust URLs to GitHub account due to renaming @munterfinger to @munterfi.
* Note: zeitsprung.fm has moved to geschichte.fm, therefore this project no longer is maintained.

0.1.0 (2020-09-22)
------------------

* First release on PyPI.
* Scraper class to download the meta data and audio files of all episodes.
* Database class to setup and access the SQLite database containing the meta data of the episodes.
* Documentation using readthedocs: https://zeitsprung.readthedocs.io/en/latest/
* Github action for building and testing the package.
* Coverage tests using codecov.io.


tests/__init__.py
"""Unit test package for zeitsprung."""


tests/test_zeitsprung.py
#!/usr/bin/env python

"""Tests for `zeitsprung` package."""

from zeitsprung.database import SQLiteEngine
from zeitsprung.scraping import Scraper


def test_database(tmp_path):
    d = tmp_path / 'data'
    d.mkdir()
    db_file = d / 'zeitsprung.db'
    print(db_file)
    db = SQLiteEngine(db_file)
    assert type(db) is SQLiteEngine


def test_scraper(tmp_path):
    d = tmp_path / 'data'
    d.mkdir()
    db_file = d / 'zeitsprung.db'
    print(db_file)
    s = Scraper(db_file)
    assert type(s) is Scraper


zeitsprung/scrape.py
from bs4 import BeautifulSoup
from datetime import datetime, timezone
from io import BytesIO
from json import loads
from pathlib import Path
from pydub import AudioSegment
from requests import get
from time import sleep
from typing import Union
from zeitsprung.base import Base
from zeitsprung.database import SQLiteEngine


class Scraper(Base):
    """Class for scraping and preprocessing the data from 'www.zeitsprung.fm'."""

    def __init__(self, data_folder: str, update_interval: int = 24*60*60,
                 reset: bool = False, verbose: bool = True) -> None:
        """
        Class constructor for the Scraper class.

        Parameters
        ----------
        data_folder : str
            Folder to store the database and audio files. Is created or if existing, the files will bind to this.
        update_interval : int, default 24*60*60
            Interval to wait for updating after the last episode is fetched.
        reset : bool, default False
            Ignore and reset an existing database.?
        verbose : bool, default True
            Print messages about the activities conducted by a class instance.

        """

        super().__init__(verbose)
        self.created_at = datetime.now(timezone.utc).replace(microsecond=0).isoformat()
        self.data_folder = Path(data_folder)
        self.db = SQLiteEngine(self.data_folder / 'zeitsprung.db')
        self.update_interval = update_interval
        self.verbose = verbose

        if (self.data_folder / 'zeitsprung.db').exists() and reset:
            self._print(f"Overwriting existing directory structure in '{data_folder}'.")
            Path(data_folder).mkdir(parents=True, exist_ok=True)
            (Path(data_folder) / 'audio').mkdir(parents=True, exist_ok=True)
            self.db.setup_schema()

        elif (self.data_folder / 'zeitsprung.db').exists() and not reset:
            self._print(f"Binding to existing directory structure in '{data_folder}'.")

        else:
            self._print(f"Creating directory structure in '{data_folder}'.")
            Path(data_folder).mkdir(parents=True, exist_ok=True)
            (Path(data_folder) / 'audio').mkdir(parents=True, exist_ok=True)
            self.db.setup_schema()

        self.current_episode = self.db.query_last_episode_id()

    def __str__(self):
        return f"Scraper created at '{self.created_at}' with db connection to " \
               f"'{self.db.db_file}', current episode is 'ZS{self.current_episode}'."

    def get_episode_meta(self, i: int) -> Union[list, None]:
        """
        Gets the episodes meta data (title, description, publication and modified at date) and stores it to the
        database.

        Parameters
        ----------
        i : int
            Number of the episode (i - 1).

        Returns
        -------
        list:
            List containing the meta data of the episode.

        """
        url = f"https://www.zeitsprung.fm/podcast/zs{'0'+str(i) if i < 10 else str(i)}/"
        self._print(f'Requesting meta data of episode {i}: {url}')
        html_doc = get(url)
        if html_doc.status_code == 200:
            soup = BeautifulSoup(html_doc.content, 'html.parser')
            script_content = loads(soup.find("script").contents[0])
            title = soup.find('title').get_text(strip=True).split(":")
            return [
                i,
                datetime.fromisoformat(self.search_key('datePublished', script_content['@graph'])),
                datetime.fromisoformat(self.search_key('dateModified', script_content['@graph'])),
                title[0],
                title[1][1:],
                soup.find("meta", {"property": "og:description"}).get('content'),
                soup.find("meta", {"property": "og:url"}).get('content'),
                None if soup.find("ul", {"class": "episode_download_list"}) is None else soup.find(
                    "ul", {"class": "episode_download_list"}).find_all('a')[0].get('href')
            ]
        else:  # html_doc.status_code == 404:
            return None

    @staticmethod
    def search_key(key, dict_obj):
        for entry in dict_obj:
            if key in entry:
                return entry[key]

    def get_episode_audio(self, url: str) -> Union[AudioSegment, None]:
        """
        Downloads the audio of a specified episode.

        Parameters
        ----------
        url : str
            URL to download the audio from.

        Returns
        -------
        AudioSegment:
            The audio of the episode.

        """
        if url is not None:
            self._print(f"Fetching audio file from {url}")
            audio_mp3 = BytesIO(get(url, allow_redirects=True).content)
            audio = AudioSegment.from_file(audio_mp3)
            return audio
        else:
            self._print(f'No audio file available for this episode.')
            return None

    def save_episode_audio(self, audio: AudioSegment, file_name: str) -> None:
        """
        Save the audio file of an episode and as '.wav' file.

        Parameters
        ----------
        audio :  AudioSegment
            Audio file to save.
        file_name : str
            File name with path, where the file should be saved to.

        Returns
        -------
        None

        """
        self._print(f"Exporting audio sequence to file '{file_name}'")
        audio.export(file_name, format="wav")

    def run(self) -> None:
        """
        Start the scraper, which will download the meta data and audio files of all not yet existing episodes in the
        database.

        Returns
        -------
        None

        """
        while True:
            meta_row = self.get_episode_meta(self.current_episode + 1)
            if meta_row is not None:
                self.db.insert_meta_row(meta_row)
                audio = self.get_episode_audio(meta_row[7])
                if audio is not None:
                    audio_row = [
                        self.current_episode + 1,
                        self.data_folder / 'audio' / f'{str(self.current_episode + 1).zfill(3)}.wav',
                        round(audio.duration_seconds),
                        audio.frame_rate,
                        audio.frame_width
                    ]
                    self.save_episode_audio(audio, audio_row[1])
                    self.db.insert_audio_row(audio_row)
                self.current_episode += 1
            else:
                self._print(f"Episode not yet published, pausing for {int(self.update_interval/(60*60))} hours.")
                sleep(self.update_interval)


zeitsprung/database.py
from pandas import DataFrame, to_datetime
from typing import Union
from sqlite3 import connect, Connection
from unicodedata import normalize
from zeitsprung.base import Base


class SQLiteEngine(Base):
    """Class to set up and access a SQLite database to store the data from the 'zeitsprung.fm' website."""

    def __init__(self, db_file: str, verbose: bool = True) -> None:
        """
        Class constructor for the Scraper class.

        Parameters
        ----------
        db_file : str
            Path to the 'zeitsprung.db' file.
            Hint: The database is created inside the data folder, specified in the constructor of 'Scrape'.
        verbose : bool, default True
            Print messages about the activities conducted by a class instance.

        Returns
        -------
        None

        """
        super().__init__(verbose)
        self.db_file = db_file
        self.verbose = verbose

    def __str__(self) -> str:
        """
        Print function of the class.

        Returns
        -------
        str
            A string, which describes the class instance.

        """
        return f"Database engine connecting to '{self.db_file}'."

    def create_connection(self) -> Connection:
        """
        Creates a connection object using the provided credentials in instance creation.

        Returns
        -------
        Connection
            A object for connecting to the database.

        """
        conn = None
        try:
            conn = connect(self.db_file)
        except BaseException as e:
            print(e)
        return conn

    def setup_schema(self) -> None:
        """
        Creates the schema with 'meta' and 'audio' relations in the 'zeitsprung.db' database.

        Returns
        -------
        None

        """
        self._print(f"Setting up SQLite database at '{self.db_file}'.")
        conn = self.create_connection()
        cur = conn.cursor()
        cur.execute('DROP TABLE IF EXISTS meta;')
        cur.execute('''
        CREATE TABLE meta (
            uid INTEGER PRIMARY KEY,
            published_at DATETIME NOT NULL,
            modified_at DATETIME NOT NULL,
            abbreviation TEXT NOT NULL,
            title TEXT NOT NULL,
            description TEXT NOT NULL,
            url_episode TEXT NOT NULL,
            url_audio TEXT NOT NULL
        );
        ''')
        cur.execute('DROP TABLE IF EXISTS audio;')
        cur.execute('''
        CREATE TABLE audio (
            uid INTEGER PRIMARY KEY,
            file_path TEXT NOT NULL,
            duration INTEGER NOT NULL,
            frame_rate INTEGER NOT NULL,
            frame_width INTEGER NOT NULL
        );
        ''')
        conn.commit()
        conn.close()

    def insert_meta_row(self, row: list) -> None:
        """
        Inserts a record row into the 'meta' relation.

        Parameters
        ----------
        row : list
            Values to insert.

        Returns
        -------
        None

        """
        self._print(f"Writing row for '{row[0]}' to table 'meta'.")
        conn = self.create_connection()
        cur = conn.cursor()
        cur.execute(f"""
        INSERT INTO meta (uid, published_at, modified_at, abbreviation, title, description, url_episode, url_audio)
        VALUES('{row[0]}', '{row[1]}', '{row[2]}', '{row[3]}', '{normalize("NFKD", row[4]).replace("'", '')}',
            '{normalize("NFKD", row[5]).replace("'", '')}', '{row[6]}', '{row[7]}');
        """)
        conn.commit()
        conn.close()

    def insert_audio_row(self, row: list) -> None:
        """
        Inserts a record row into the 'audio' relation.

        Parameters
        ----------
        row : list
            Values to insert.

        Returns
        -------
        None

        """
        self._print(f"Writing row for '{row[0]}' to table 'audio'.")
        conn = self.create_connection()
        cur = conn.cursor()
        cur.execute(f"""
        INSERT INTO audio (uid, file_path, duration, frame_rate, frame_width)
        VALUES('{row[0]}', '{row[1]}', '{row[2]}', '{row[3]}', '{row[4]}');
        """)
        conn.commit()
        conn.close()

    def query_last_episode_id(self) -> Union[int, None]:
        """
        Queries the last episode from the 'meta' relation.

        Returns
        -------
        int or None
            The id of the last episode.

        """
        conn = self.create_connection()
        cur = conn.cursor()
        cur.execute('SELECT max(uid) FROM meta')
        uid = cur.fetchall()[0][0]
        return 0 if uid is None else uid

    def query_all_meta(self) -> DataFrame:
        """
        Queries the complete record from the 'meta' relation and parses it into a DataFrame object.

        Returns
        -------
        DataFrame
            The meta table.

        """
        self._print("Querying all rows from table 'meta'.")
        conn = self.create_connection()
        cur = conn.cursor()
        cur.execute('SELECT * FROM meta')
        rows = cur.fetchall()
        conn.close()
        df = DataFrame([list(row) for row in rows],
                       columns=['uid', 'published_at', 'modified_at', 'abbreviation',
                                'title', 'description', 'url_episode', 'url_audio'])
        df['published_at'] = to_datetime(df['published_at'])
        df['modified_at'] = to_datetime(df['modified_at'])
        return df

    def query_all_audio(self) -> DataFrame:
        """
        Queries the complete record from the 'audio' relation and parses it into a DataFrame object.

        Returns
        -------
        DataFrame
            The audio table.

        """
        self._print("Querying all rows from table 'audio'.")
        conn = self.create_connection()
        cur = conn.cursor()
        cur.execute('SELECT * FROM audio')
        rows = cur.fetchall()
        conn.close()
        df = DataFrame([list(row) for row in rows],
                       columns=['uid', 'file_path', 'duration', 'frame_rate', 'frame_width'])
        return df


zeitsprung/scraping.py
from bs4 import BeautifulSoup
from datetime import datetime, timezone
from io import BytesIO
from json import loads
from pathlib import Path
from pydub import AudioSegment
from requests import get
from time import sleep
from typing import Union
from zeitsprung.base import Base
from zeitsprung.database import SQLiteEngine


class Scraper(Base):
    """Class for scraping and preprocessing the data from the 'www.zeitsprung.fm' website."""

    def __init__(self, data_folder: str, update_interval: int = 24*60*60,
                 reset: bool = False, verbose: bool = True) -> None:
        """
        Class constructor for the Scraper class.

        Parameters
        ----------
        data_folder : str
            Folder to store the database and audio files. Is created or if existing, the files will bind to this.
        update_interval : int, default 24*60*60
            Interval to wait for updating after the last episode is fetched.
        reset : bool, default False
            Ignore and reset an existing database.?
        verbose : bool, default True
            Print messages about the activities conducted by a class instance.

        """

        super().__init__(verbose)
        self.created_at = datetime.now(timezone.utc).replace(microsecond=0).isoformat()
        self.data_folder = Path(data_folder)
        self.db = SQLiteEngine(self.data_folder / 'zeitsprung.db')
        self.update_interval = update_interval
        self.verbose = verbose

        if (self.data_folder / 'zeitsprung.db').exists() and reset:
            self._print(f"Overwriting existing directory structure in '{data_folder}'.")
            Path(data_folder).mkdir(parents=True, exist_ok=True)
            (Path(data_folder) / 'audio').mkdir(parents=True, exist_ok=True)
            self.db.setup_schema()

        elif (self.data_folder / 'zeitsprung.db').exists() and not reset:
            self._print(f"Binding to existing directory structure in '{data_folder}'.")

        else:
            self._print(f"Creating directory structure in '{data_folder}'.")
            Path(data_folder).mkdir(parents=True, exist_ok=True)
            (Path(data_folder) / 'audio').mkdir(parents=True, exist_ok=True)
            self.db.setup_schema()

        self.current_episode = self.db.query_last_episode_id()

    def __str__(self) -> str:
        """
        Print function of the class.

        Returns
        -------
        str
            A string, which describes the class instance.

        """
        return f"Scraper created at '{self.created_at}' with db connection to " \
               f"'{self.db.db_file}', current episode is 'ZS{self.current_episode}'."

    def get_episode_meta(self, i: int) -> Union[list, None]:
        """
        Gets the episodes meta data (title, description, publication and modified at date) and stores it to the
        database.

        Parameters
        ----------
        i : int
            Number of the episode (i - 1).

        Returns
        -------
        list:
            List containing the meta data of the episode.

        """
        url = f"https://www.zeitsprung.fm/podcast/zs{'0'+str(i) if i < 10 else str(i)}/"
        self._print(f'Requesting meta data of episode {i}: {url}')
        html_doc = get(url)
        if html_doc.status_code == 200:
            soup = BeautifulSoup(html_doc.content, 'html.parser')
            script_content = loads(soup.find("script").contents[0])
            title = soup.find('title').get_text(strip=True).split(":")
            return [
                i,
                datetime.fromisoformat(self.search_key('datePublished', script_content['@graph'])),
                datetime.fromisoformat(self.search_key('dateModified', script_content['@graph'])),
                title[0],
                title[1][1:],
                soup.find("meta", {"property": "og:description"}).get('content'),
                soup.find("meta", {"property": "og:url"}).get('content'),
                None if soup.find("ul", {"class": "episode_download_list"}) is None else soup.find(
                    "ul", {"class": "episode_download_list"}).find_all('a')[0].get('href')
            ]
        else:  # html_doc.status_code == 404:
            return None

    @staticmethod
    def search_key(key, dict_obj):
        for entry in dict_obj:
            if key in entry:
                return entry[key]

    def get_episode_audio(self, url: str) -> Union[AudioSegment, None]:
        """
        Downloads the audio of a specified episode.

        Parameters
        ----------
        url : str
            URL to download the audio from.

        Returns
        -------
        AudioSegment:
            The audio of the episode.

        """
        if url is not None:
            self._print(f"Fetching audio file from {url}")
            audio_mp3 = BytesIO(get(url, allow_redirects=True).content)
            audio = AudioSegment.from_file(audio_mp3)
            return audio
        else:
            self._print('No audio file available for this episode.')
            return None

    def save_episode_audio(self, audio: AudioSegment, file_name: str) -> None:
        """
        Save the audio file of an episode and as '.wav' file.

        Parameters
        ----------
        audio :  AudioSegment
            Audio file to save.
        file_name : str
            File name with path, where the file should be saved to.

        Returns
        -------
        None

        """
        self._print(f"Exporting audio sequence to file '{file_name}'")
        audio.export(file_name, format="wav")

    def run(self) -> None:
        """
        Start the scraper, which will download the meta data and audio files of all not yet existing episodes in the
        database.

        Returns
        -------
        None

        """
        while True:
            meta_row = self.get_episode_meta(self.current_episode + 1)
            if meta_row is not None:
                self.db.insert_meta_row(meta_row)
                audio = self.get_episode_audio(meta_row[7])
                if audio is not None:
                    audio_row = [
                        self.current_episode + 1,
                        self.data_folder / 'audio' / f'{str(self.current_episode + 1).zfill(3)}.wav',
                        round(audio.duration_seconds),
                        audio.frame_rate,
                        audio.frame_width
                    ]
                    self.save_episode_audio(audio, audio_row[1])
                    self.db.insert_audio_row(audio_row)
                self.current_episode += 1
            else:
                self._print(f"Episode not yet published, pausing for {int(self.update_interval/(60*60))} hours.")
                sleep(self.update_interval)
